components/custom_board/ma120x0/MerusAudio.c

```diff
@@ -51,48 +49,75 @@ static const char* I2C_TAG = "i2c";
 #define NACK_VAL   0x1         /*!< I2C nack value */
 
 
-void setup_ma120x0()
-{  // Setup control pins nEnable and nMute
-   gpio_config_t io_conf;
-
-   io_conf.intr_type = GPIO_PIN_INTR_DISABLE;
-   io_conf.mode = GPIO_MODE_OUTPUT;
-   io_conf.pin_bit_mask = (1ULL<<MA_NENABLE_IO | 1ULL<<MA_NMUTE_IO);
-   io_conf.pull_down_en = 0;
-   io_conf.pull_up_en = 0;
-
-   gpio_config(&io_conf);
-
-   gpio_set_level(MA_NMUTE_IO, 0);
-   gpio_set_level(MA_NENABLE_IO, 1);
+static i2c_config_t i2c_cfg;
 
-   i2c_master_init();
+audio_hal_func_t AUDIO_CODEC_MA120X0_DEFAULT_HANDLE = {
+    .audio_codec_initialize = ma120x0_init,
+    .audio_codec_deinitialize = ma120x0_deinit,
+    .audio_codec_ctrl = ma120x0_ctrl,
+    .audio_codec_config_iface = ma120x0_config_iface,
+    .audio_codec_set_mute = ma120x0_set_mute,
+    .audio_codec_set_volume = ma120x0_set_volume,
+    .audio_codec_get_volume = ma120x0_get_volume,
+    .audio_hal_lock = NULL,
+    .handle = NULL,
+};
 
-   gpio_set_level(MA_NENABLE_IO, 0);
+esp_err_t ma120x0_deinit(void)
+{
+    // TODO
+    return ESP_OK;
+}
 
-   uint8_t res = ma_read_byte(MA120X0_ADDR,1,MA_hw_version__a);
-   printf("Hardware version: 0x%02x\n",res);
+esp_err_t ma120x0_ctrl(audio_hal_codec_mode_t mode, audio_hal_ctrl_t ctrl_state)
+{
+    // TODO
+    return ESP_OK;
+}
 
-   ma_write_byte(MA120X0_ADDR,1,MA_i2s_format__a,8);          // Set i2s left justified, set audio_proc_enable
-   ma_write_byte(MA120X0_ADDR,1,MA_vol_db_master__a,0x50);    // Set vol_db_master
+esp_err_t ma120x0_config_iface(audio_hal_codec_mode_t mode, audio_hal_codec_i2s_iface_t *iface)
+{
+    //TODO
+    return ESP_OK;
+}
 
-   res = ma_read_byte(MA120X0_ADDR,1,MA_error__a);
-   printf("Errors : 0x%02x\n",res);
+esp_err_t ma120x0_set_volume(int vol)
+{
+	esp_err_t ret = ESP_OK;
+	uint8_t cmd[4];
+	cmd[0] = 128-vol;
+	cmd[1] = cmd[0];
+	ma_write(0x20, 2, 0x0003, cmd, 2);
+	return ret;
+}
 
-   res = 01; // get_MA_audio_in_mode_mon();
-   printf("Audio in mode : 0x%02x\n",res);
+esp_err_t ma120x0_get_volume(int *vol)
+{
+	esp_err_t ret = ESP_OK;
+    uint8_t rxbuf[12];
+    ma_read(0x20, 2, 0x0003, rxbuf, 3);
+    // printf("\nVolume : 0x%02x 0x%02x 0x%02x\n", rxbuf[0], rxbuf[1], rxbuf[2]);
+	*vol = rxbuf[0];  // XXX which register? also 128+/-?
```

Comment:
Note the XXX here, not sure the correct implementation for this function.

Reply:
Need conversion here - Read out from 2 volume register is needed or must be synced. Adr 0x0003 ch0 0x0004 ch1. 
0x18 is 0 dB gain on volume and and each step up 0.75 dB attenuation - Amplifier has 20 dB build in. 
Just let me take over from here.